homeassistant/components/shelly/__init__.py

```diff
@@ -142,6 +146,60 @@ async def async_device_setup(
         )
 
 
+async def async_services_setup(
+    hass: HomeAssistant, dev_reg: device_registry.DeviceRegistry
+):
+    """Set up services."""
+
+    async def async_service_ota_update(call: ServiceCall):
+        """Trigger OTA update."""
+        if not (call.data.get(ATTR_DEVICE_ID) or call.data.get(ATTR_AREA_ID)):
+            _LOGGER.warning("OTA update service: no device or area selected")
+            return
+
+        devices = []
+        if call.data.get(ATTR_AREA_ID):
+            for area_id in call.data.get(ATTR_AREA_ID):
+                devices += [
+                    area_dev
+                    for area_dev in device_registry.async_entries_for_area(
+                        dev_reg, area_id
+                    )
+                    if DOMAIN in next(iter(area_dev.identifiers))
+                ]
+
+        if call.data.get(ATTR_DEVICE_ID):
+            for device_id in call.data.get(ATTR_DEVICE_ID):
+                device = dev_reg.async_get(device_id)
+                if not any(device.id == x.id for x in devices):
+                    devices += [device]
+
+        for device in devices:
+            entry_id = next(iter(device.config_entries))
+            entry_data = hass.data[DOMAIN][DATA_CONFIG_ENTRY][entry_id]
+            rest: ShellyDeviceRestWrapper = entry_data[REST]
+            update_data = rest.device.status["update"]
+
+            if not update_data["has_update"]:
+                _LOGGER.info("No OTA updates for %s available", device.name)
+                continue
+
+            if update_data["status"] == "updating":
+                _LOGGER.warning("OTA update already in progress for %s", device.name)
+                continue
+
+            _LOGGER.debug(
+                "OTA update service - trigger OTA update for device %s from '%s' to '%s'",
+                device.name,
+                update_data["old_version"],
+                update_data["new_version"],
+            )
+            resp = await rest.device.http_request("get", "ota", {"update": "true"})
+            _LOGGER.debug("OTA update service - response: %s", resp)
+
+    hass.services.async_register(DOMAIN, SERVICE_OTA_UPDATE, async_service_ota_update)
```

Comment:
This should be implemented as an entity service:

<https://developers.home-assistant.io/docs/dev_101_services#entity-services>

Reply:
Hi @frenck 
sorry but I'm not sure if I get it correct :confused: 
The service I want to implement is not related to a specific entity/sensor or platform, but to a device - therefore I added this service as integration service.
To be honest I was also wondering why I have to extract the device_ids from area_id (_and filter them to get only shelly devices_) to get afterwards the (_config\__)entry_id  from the devices and in addition have to care about that no duplicated device_ids were processed :thinking: 
Surely it is more realistic that I'm total wrong and have missed something essential :see_no_evil: 
I would be very grateful for hints on this, also about extended examples.

Reply:
When implemented as an entity service, it will support a `target`, in that case Home Assistant will resolve/handle devices, areas and entity ID references for you, and call the matching entity services (no matter what the user has put in).